### PR TITLE
Refactor instrumentation

### PR DESCRIPTION
**Pull request by dstepanov**
`InvocationInstrumenter` is only capturing context
`Instrumentation` is setting/removing context using try-catch-resources

@lgathy @graemerocher WDYT?

### CODE REVIEW

**Comment by graemerocher** [comment]
From an initial look seems much cleaner, although lets see if the tests pass and need to test it against #3427 on `master`

This test should probably not be removed but rather adapted to the new API no? https://github.com/micronaut-projects/micronaut-core/pull/3440/files#diff-cce7e75be549c3bb4520519b0c967f38L40

**Comment by dstepanov** [comment]
What should be adapted? The link doesn't work

**Comment by graemerocher** [comment]
`RxChainedInstrumentationSpec.groovy`

**Comment by graemerocher** [comment]
@lgathy can speak better for that test I imagine

**Comment by dstepanov** [comment]
`InvocationInstrumenter` cannot be called recursively anymore, so it's not needed.
 
@graemerocher Looks like discovery client failed, maybe it can be extracted from the core in the future, it takes a long time and fails for me.

**Comment by graemerocher** [comment]
@dstepanov yeah we have been meaning to extract that since Eureka's eventual consistency model makes it hard to create integration tests that don't fail occasionally

**Comment by dstepanov** [comment]
@lgathy Would you have time to review this changes?

**Comment by lgathy** [comment]
@dstepanov Sure, I can do it tonight if that's OK.

**Comment by lgathy** [comment]
@dstepanov Regarding how to port this back to `1.x`: Is it a viable option to add the new Instrumentation API while also keeping before/afterInvocation methods (marking them deprecated) + re-adding the `DefaultInstrumentation` returned in the default impementation of `InvocationInstrumenter.newInstrumentation()`? This will make 3rd party instrumenter implementations non-thread-safe, but it will prevent them from not compiling and make them work correctly until they are being reused concurrently, which I would still consider a fairly rare use-case.

@graemerocher Would you feel comfortible having this new Instrumentation API added in `1.3.x`? Technically it is a new public API, which would require a new minor version. Is there a plan for `1.4.0` before `2.0.0` at all?

Btw, if we choose to do this, we might consider adjusting the `@since 2.0` markers for the sake of correctness in the new files here (in master), too.

File: tracing/src/main/java/io/micronaut/tracing/instrument/util/MdcInstrumenter.java [hunk 1]
```diff
@@ -47,23 +49,24 @@ public InvocationInstrumenter newInvocationInstrumenter() {
...
-
+            @NonNull
             @Override
-            public void beforeInvocation() {
-                oldContextMap = MDC.getCopyOfContextMap();
+            public Instrumentation newInstrumentation() {
+                Map<String, String> oldContextMap = MDC.getCopyOfContextMap();
                 if (contextMap != null && !contextMap.isEmpty()) {
                     MDC.setContextMap(contextMap);
                 }
```
[the remark this answers is not in the document]
Comment (dstepanov) [reply tracing/src/main/java/io/micronaut/tracing/instrument/util/MdcInstrumenter.java:58]: There can be multiple MdcIstrumenters and `clear` would make the previous context values disappear. There is a test that actually is testing it.

File: http-server/src/main/java/io/micronaut/http/server/context/ServerRequestContextInvocationInstrumenter.java [hunk 2]
```diff
@@ -28,40 +30,40 @@
...
-     */
-    @Override
-    public void afterInvocation(boolean cleanup) {
-        if (isSet || cleanup) {
-            ServerRequestContext.set(cleanup ? null : currentRequest);
+        } else {
             isSet = false;
         }
-    }
+        return new Instrumentation() {
```
[the remark this answers is not in the document]
Comment (dstepanov) [reply http-server/src/main/java/io/micronaut/http/server/context/ServerRequestContextInvocationInstrumenter.java:56]: I don't think there is a big perf. difference.

**Comment by dstepanov** [comment]
I think `beforeInvocation()` and `afterInvocation()` can be removed in 2.0.
@graemerocher These changes should be probably ported to RxJava3 lib, is it supposed to be 1.3 compatible?

File: http-server/src/main/java/io/micronaut/http/server/context/ServerRequestContextInvocationInstrumenter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (SplotyCode) [reply http-server/src/main/java/io/micronaut/http/server/context/ServerRequestContextInvocationInstrumenter.java:56]: At Link time Lambda are slower after that there behave very similar to anonymous classes. Lambdas only have instance variables if required and stateless lambdas get optimised but other then that there is not mutch different.

**Comment by graemerocher** [comment]
No currently it is rxjava 3 only https://github.com/micronaut-projects/micronaut-rxjava3/blob/master/gradle.properties#L4

**Comment by graemerocher** [comment]
@lgathy There are no plans right now for a 1.4.x as 2.0.0 is going to be released this month, a backport is not out of the question, but it would need to be non breaking in nature. Now given the instrumentation APIs are broken anyway maybe that is not a big deal.

For 2.0 I will merge this but I am going to mark all the new APIs with `@Internal` and `@Experimental` in case changes need to be made post release.

**Comment by graemerocher** [comment]
Thanks to both of your for the contributions

**Pull request merged**